### ISSUE TITLE
Auth error only after recaptcha: "Cannot read property 'style' of null"

### ISSUE THREAD

**Issue by zachasme** [firebase/firebase-js-sdk#3356]
### Describe your environment

  * Operating System version: Linux 5.7.5-arch1-1 x86_64
  * Browser version: Firefox 79.0b3 (64-bit)
  * Firebase SDK version: 7.15.5
  * Firebase Product: auth

### Describe the problem

I am setting up Phone Auth with firebase. If there is a problem with the phone number (too long, too short, etc), but the captcha is shown (delaying the error until after resolving the captcha), the following error is thrown:

```
Uncaught TypeError: Cannot read property 'style' of null
```

I have a feeling it happens when firebase auth attemps to reset the recaptcha after a problem with phone auth occurs.

There is no issue when captcha is shown but phone number is valid. 

#### Steps to reproduce:

* Go to the below codesandbox (open it in incognito -- hopefully that'll make sure captcha is shown)
* Open console
* Click submit (with an invalid phone number)
* Ensure captcha is shown, if not, refresh and try again until it is
* Observe the following error is thrown:

```
Uncaught TypeError: can't access property "style", r is null
    B https://www.gstatic.com/recaptcha/releases/NMoy4HgGiLr5NAQaEQa2ho8X/recaptcha__en.js:310
    Y https://www.gstatic.com/recaptcha/releases/NMoy4HgGiLr5NAQaEQa2ho8X/recaptcha__en.js:120
    setTimeout handler*n</< https://www.gstatic.com/recaptcha/releases/NMoy4HgGiLr5NAQaEQa2ho8X/recaptcha__en.js:132
    L https://www.gstatic.com/recaptcha/releases/NMoy4HgGiLr5NAQaEQa2ho8X/recaptcha__en.js:120
    L https://www.gstatic.com/recaptcha/releases/NMoy4HgGiLr5NAQaEQa2ho8X/recaptcha__en.js:123
    B https://www.gstatic.com/recaptcha/releases/NMoy4HgGiLr5NAQaEQa2ho8X/recaptcha__en.js:300
    <anonymous> https://www.gstatic.com/recaptcha/releases/NMoy4HgGiLr5NAQaEQa2ho8X/recaptcha__en.js:654
    T https://www.gstatic.com/recaptcha/releases/NMoy4HgGiLr5NAQaEQa2ho8X/recaptcha__en.js:125
    T https://www.gstatic.com/recaptcha/releases/NMoy4HgGiLr5NAQaEQa2ho8X/recaptcha__en.js:125
    B https://www.gstatic.com/recaptcha/releases/NMoy4HgGiLr5NAQaEQa2ho8X/recaptcha__en.js:351
    next https://www.gstatic.com/recaptcha/releases/NMoy4HgGiLr5NAQaEQa2ho8X/recaptcha__en.js:107
    u https://www.gstatic.com/recaptcha/releases/NMoy4HgGiLr5NAQaEQa2ho8X/recaptcha__en.js:266
    promise callback*C https://www.gstatic.com/recaptcha/releases/NMoy4HgGiLr5NAQaEQa2ho8X/recaptcha__en.js:266
    N https://www.gstatic.com/recaptcha/releases/NMoy4HgGiLr5NAQaEQa2ho8X/recaptcha__en.js:266
    f https://www.gstatic.com/recaptcha/releases/NMoy4HgGiLr5NAQaEQa2ho8X/recaptcha__en.js:266
    f https://www.gstatic.com/recaptcha/releases/NMoy4HgGiLr5NAQaEQa2ho8X/recaptcha__en.js:218
    L https://www.gstatic.com/recaptcha/releases/NMoy4HgGiLr5NAQaEQa2ho8X/recaptcha__en.js:125
    JA https://www.gstatic.com/recaptcha/releases/NMoy4HgGiLr5NAQaEQa2ho8X/recaptcha__en.js:369
    G https://www.gstatic.com/recaptcha/releases/NMoy4HgGiLr5NAQaEQa2ho8X/recaptcha__en.js:53
    vE https://www.gstatic.com/recaptcha/releases/NMoy4HgGiLr5NAQaEQa2ho8X/recaptcha__en.js:449
    E https://www.gstatic.com/recaptcha/releases/NMoy4HgGiLr5NAQaEQa2ho8X/recaptcha__en.js:26
recaptcha__en.js:310:363
```
#### Relevant Code:

https://codesandbox.io/s/compassionate-bouman-vw5zn?file=/index.html

#### Related issues:

* https://github.com/dozoisch/react-google-recaptcha/issues/129
* https://github.com/google/google-api-javascript-client/issues/281

**Comment by MeghaB** [comment]
Tracking internally via b/161141499

Thanks @zachasme for the repro steps here! We'll update back when we know more.

**Comment by vdjurdjevic** [comment]
Any workaround for this? My application has only phone auth, this makes it unusable..

**Comment by MeghaB** [comment]
Are you using Firebase UI (or lib-phonenumber, or another client-side phone verification) to ensure the phone number is valid before triggering the reCAPTCHA flow?

**Comment by juliosmax** [comment]
i am using firebase_auth: "^0.18.1+2" for flutterFire and i got the same issue: TypeError: Cannot read property 'style' of null
    at Array.<anonymous> (https://www.gstatic.com/recaptcha/releases/48TunWH-ZrLteSwFVbw6tVnx/recaptcha__es.js:43:479)
    at eH.<anonymous> (https://www.gstatic.com/recaptcha/releases/48TunWH-ZrLteSwFVbw6tVnx/recaptcha__es.js:66:353)

**Comment by zachasme** [comment]
@MeghaB no I am only using the reCAPTCHA and Firebase JS SDK. Following the [phone auth docs](https://firebase.google.com/docs/auth/web/phone-auth) this is the simplest implementation I could come up with.

```js
const recaptchaRef = document.getElementById("recaptcha"); // this is a button
const phoneNumberRef = document.getElementById("phoneNumber"); // this is an input

async function handleSignIn(verifier) {
  try {
    const { verificationId } = await firebase
      .auth()
      .signInWithPhoneNumber(phoneNumberRef.value, verifier);
    console.log(verificationId);
  } catch (error) {
    console.error(error);
  }
}

const verifier = new firebase.auth.RecaptchaVerifier(recaptchaRef, {
  size: "invisible",
  callback: () => handleSignIn(verifier)
});

verifier.render();
```

This is also what runs in the [reproduction sandbox](https://codesandbox.io/s/compassionate-bouman-vw5zn?file=/index.html).

If it is simply a requirement that library-consumers need validate phone numbers before calling `signInWithPhoneNumber` then I suggest it should be part of the phone auth docs linked above.

@yuchenshi I will be happy to provide any info you would like :+1:

**Comment by bumbleirl** [comment]
getting the same error. any updates?

**Comment by mrtk0y** [comment]
Still got this error with reCaptcha v2, using with Firebase, and 
```
https://www.npmjs.com/package/react-google-recaptcha
```
![https://i.gyazo.com/9971295c05f6daa4cd85e19c996f215d.gif](https://i.gyazo.com/9971295c05f6daa4cd85e19c996f215d.gif)

**Comment by kanafghan** [comment]
I also get this error using `@angular/fire` v5 and Angular 9. Any updates on this?

**Comment by bgordon01** [comment]
I'm also getting this error using @angular/cdk & @angular/material v10.2.7 and firebase 8.0.1

**Comment by zachrnolan** [comment]
I'm also getting this error using Firebase JS SDK, any updates?

**Comment by sayfulloev** [comment]
I started getting this error once I enabled `Phone numbers for testing (optional)` option in Firebase. Once I removed it and tried to login with regular phone number it worked just fine.

**Comment by aghahasaan** [comment]
It's because you are using test numbers which are not really receiving the response code.

**Comment by Collins1738** [comment]
> It's because you are using test numbers which are not really receiving the response code.

I used a real number and got the same error. I also got the verification code text too.

**Comment by aghahasaan** [comment]
Then it’s most likely you have placed the code blocks in wrong order, also, if you 
 have ‘allow users to sign in with same email from different platforms/providers’ option enabled can possibly cause this error as well. I got mine working without any errors using all these steps and by deleting all test numbers I had entered before.

**Comment by willvlad** [comment]
It's a critical bug that derails any implementation of phone auth in web apps. 

Anyone from Firebase reading this and working on a fix? 

Just disable visible captcha if the setting is invisible or don't ask for property style of null!

**Comment by Collins1738** [comment]
@willvlad While I hope this bug gets fixed soon, it can actually be ignored. The error pops up in deployment but in production it just gets logged and it doesn’t affect anything. Authentication still works as planned. That has been my experience with it.

**Comment by maitrungduc1410** [comment]
I'm still facing this problem

**Comment by YouZag** [comment]
Still a problem when the size: invisible shows the pictures to match.

**Comment by YouZag** [comment]
I'm confused - is this not a problem for everyone?

**Comment by hirashuriken** [comment]
I have the same problem for web app. Does someone know how to fix it? It looks like a critical issue!

**Comment by nicks258** [comment]
Any updates?

**Comment by ga0321** [comment]
push

**Comment by jianweic** [comment]
I have the same problem. Does someone know how to fix it?

**Comment by olehreznichenko** [comment]
Same with me 😢

**Comment by joaovictormilhomem** [comment]
I had this problem, and I resolved by activating the authentication method in question in the firebase console, I had forgotten to do this.

**Comment by iankitverma** [comment]
Any updates?

**Comment by iankitverma** [comment]
How did you resolve it?

> I had this problem, and I resolved by activating the authentication method in question in the firebase console, I had forgotten to do this.

**Comment by landsman** [comment]
Exactly problem, on localhost with testing numbers.
Any news? @MeghaB?

> Still got this error with reCaptcha v2, using with Firebase, and
> 
> ```
> https://www.npmjs.com/package/react-google-recaptcha
> ```
> 
> ![https://i.gyazo.com/9971295c05f6daa4cd85e19c996f215d.gif](https://camo.githubusercontent.com/92a3c23ba053343c5b216c6a465324d9a18b1dfdfa8e25ffc3c1c5888588efa0/68747470733a2f2f692e6779617a6f2e636f6d2f39393731323935633035663664616134636438356531396339393666323135642e676966)

**Comment by Mayank-Sh07** [comment]
If you're using reCaptcha with a project that already has Firebase, make sure to go to the [firebase console](https://console.firebase.google.com/) and enable `authentication` for that particular project. Hope this helps

**Comment by cedvdb** [comment]
I'm having this issue with a test phone number on flutter when hot reloading.

**Comment by babakhoo** [comment]
I found this topic because I have the same error.
I found a way to fix it and go ahead while its not resolved yet.
I really hope firebase team do something about it while Im a bit disappointed that after all this time there is no move.
My solution to go ahead :
I dont use webpack for this part of program, it means I have :
import {getAuth, RecaptchaVerifier } from 'https://www.gstatic.com/firebasejs/9.6.10/firebase-auth.js';
Then without debuging, with chrom web server I host my src folder (index.html and index.js reference as 'module' ) in local host and boom No error Mrs.Captcha show out beautifully and do the job , while if I debug the same code I see internal error, and if I make main.js and bundler stuff, the same problem when 
window.recaptchaVerifier = new RecaptchaVerifier

**Comment by ritualjoint** [comment]
> but in production it just gets logged and it doesn’t affect anything. Authentication still works as planned. That has been my experience with it.

**Comment by hoangnn000** [comment]
My case was that I didn't whitelist my server domain in firebase. I just added mine into Firebase's `Authorized domains` and it's worked

**Comment by thinkingatoms** [comment]
+1 with web recaptcha.

out of curiosity how are you guys specifying the recaptcha site key in firebase flutter?  thanks in advance!

**Comment by AlkarE** [comment]
I solved this by wrapping code in try...catch

**Comment by landsman** [comment]
Guys, any progress on this?

**Comment by AlkarE** [comment]
I don't remember exactly.
I guess this part of code may give you an idea:
try {

      if(!window.recaptchaVerifier) {
  
        window.recaptchaVerifier = new firebase.auth.RecaptchaVerifier('signin-recaptcha', {
          'size': 'invisible',
          'callback': (response) => {
            // cb
          }
        });
      } 
      else {
        window.recaptchaVerifier.render().then(function(widgetId) {
          grecaptcha.reset(widgetId);
        });
      }
    } catch(e) {
      console.log(e)
    }

**Comment by alexkobylansky** [comment]
> I don't remember exactly. I guess this part of code may give you an idea: try {
> 
> ```
>   if(!window.recaptchaVerifier) {
> 
>     window.recaptchaVerifier = new firebase.auth.RecaptchaVerifier('signin-recaptcha', {
>       'size': 'invisible',
>       'callback': (response) => {
>         // cb
>       }
>     });
>   } 
>   else {
>     window.recaptchaVerifier.render().then(function(widgetId) {
>       grecaptcha.reset(widgetId);
>     });
>   }
> } catch(e) {
>   console.log(e)
> }
> ```

Sorry I don't understand where you are importing grecaptcha. Could you please guide me how do you import this? I use firebase 9.22.0

**Comment by AlkarE** [comment]
Hi.  I've used 'try...catch' for recaptcha initializing, nothing else. Why do you ask how I imported recaptcha module? you may import it as standart module. Or I missed something?

**Comment by alexkobylansky** [comment]
> Hi. I've used 'try...catch' for recaptcha initializing, nothing else. Why do you ask how I imported recaptcha module? you may import it as standart module. Or I missed something?

See I'm using React 18 and Firebase 9.22.0 version. And when I write grecaptcha.reset(), TS writes to me: "Cannot find name 'grecaptcha'". And I can't understand from where, from which module I should import

**Comment by prameshj** [comment]
I believe this error happens when there is a problem with phone auth, which in turn resets recaptcha verifier. The problem does not happen if we invoke the clear() method (as documented in https://firebase.google.com/docs/auth/web/multi-factor#enrolling_a_second_factor).

Fixing this in https://github.com/firebase/firebase-js-sdk/pull/7574, will update the PR once i can verify it with the demo app.

**Comment by prameshj** [comment]
I'm unable to repro this with the demo app in https://github.com/firebase/firebase-js-sdk/tree/master/packages/auth/demo

Are folks still seeing this issue?

**Comment by landsman** [comment]
@prameshj did you tried this with test numbers, dev mode?

**Comment by dominoFire** [comment]
I am observing this error when testing my React webapp simulating unstable connection by switching Airplane Mode on and off on my dev laptop: let the app loads, turn airplane mode on, try to sign in.

Calling the snippet below when auth fails, i.e., when `signInWithPhoneNumber()` promise fails, captured with `.catch()` call

```javascript
        signInWithPhoneNumber(auth, phoneNumber, appVerifier)
            .then((confirmationResult) => {
              // SMS sent. Prompt user to type the code from the message, then sign the
              // user in with confirmationResult.confirm(code).

            })
            .catch((error) => {
              // Error; SMS not sent
              // ...
              alert('Could not Sign In');
              console.log('Error on sending SMS: ' + error);
              if (typeof grecaptcha !== 'undefined')
              {
                console.log('resetting grecaptcha');
                grecaptcha.reset();
              }
              else
              {
                console.log('No grecaptcha found');
              }
              if (typeof window.recaptchaVerifier !== 'undefined')
              {
                console.log('resetting window.recaptchaVerifier');
                try {
                  window.recaptchaVerifier.clear();
                }
                catch
                {
                  console.log('Could not clear reCaptchaVerifier');
                }
              }
              else
              {
                console.log('No window.recaptchaVerifier found');
              }
          });
```

**Comment by prameshj** [comment]
> @prameshj did you tried this with test numbers, dev mode?

I tried with test and real phone numbers and did not see the error. I used a malformed phone number(like omitting + sign or country code) to try to trigger the error. 
Basically, i can only get the checkbox reCAPTCHA challenge.

**Comment by prameshj** [comment]
> I am observing this error when testing my React webapp simulating unstable connection by switching Airplane Mode on and off on my dev laptop: let the app loads, turn airplane mode on, try to sign in.
> 
> Calling the snippet below when auth fails, i.e., when `signInWithPhoneNumber()` promise fails, captured with `.catch()` call



You mean you see the console log of "'Error on sending SMS: Cannot read property 'style' of null"?


Does this happen when you use verifyPhoneNumber() too? This snippet is very similar to what we have in the demo app - https://github.com/firebase/firebase-js-sdk/blob/2d0a9f5fd921568bc76fcdee325b9ab5e6be8a58/packages/auth/demo/src/index.js#L722
> 
> ```js
>         signInWithPhoneNumber(auth, phoneNumber, appVerifier)
>             .then((confirmationResult) => {
>               // SMS sent. Prompt user to type the code from the message, then sign the
>               // user in with confirmationResult.confirm(code).
> 
>             })
>             .catch((error) => {
>               // Error; SMS not sent
>               // ...
>               alert('Could not Sign In');
>               console.log('Error on sending SMS: ' + error);
>               if (typeof grecaptcha !== 'undefined')
>               {
>                 // see https://firebase.google.com/docs/auth/web/phone-auth#send-a-verification-code-to-the-users-phone
>                 console.log('resetting grecaptcha');
>                 grecaptcha.reset();
>               }
>               else
>               {
>                 console.log('No grecaptcha found');
>               }
>               if (typeof window.recaptchaVerifier !== 'undefined')
>               {
>                 // see https://firebase.google.com/docs/auth/web/multi-factor#enrolling_a_second_factor
>                 console.log('resetting window.recaptchaVerifier');
>                 try {
>                   window.recaptchaVerifier.clear();
>                 }
>                 catch
>                 {
>                   console.log('Could not clear reCaptchaVerifier');
>                 }
>               }
>               else
>               {
>                 console.log('No window.recaptchaVerifier found');
>               }
>           });
> ```

**Comment by dominoFire** [comment]
> You mean you see the console log of "'Error on sending SMS: Cannot read property 'style' of null"?

No, I mean, I have the snippet from comment above in my tests to verify that I'm clearing recaptcha.

> Does this happen when you use verifyPhoneNumber() too? This snippet is very similar to what we have in the demo app -

I don't know, I see this happens in signInWithPhoneNumber. Using the debugger shows me this is the line that throws the exception

<img width="896" alt="image" src="https://github.com/firebase/firebase-js-sdk/assets/1192347/02e719ce-4c99-4266-90f3-0085c1476a19">

**Comment by zachasme** [comment]
> I'm unable to repro this with the demo app in https://github.com/firebase/firebase-js-sdk/tree/master/packages/auth/demo
> 
> Are folks still seeing this issue?

Hi @prameshj, it's been a couple of years so I've updated my original reproduction to use the latest version of the modular firebase sdk. Steps to reproduce:

- Go to the [updated codesandbox](https://codesandbox.io/s/upbeat-meadow-mxljrl?file=/src/index.mjs) (open it in incognito -- hopefully that'll make sure captcha is shown)
- Open console
- Click submit (with an invalid phone number)
- **Ensure captcha is shown**, if not, refresh and figure out a way to trigger it
- Complete the captcha
- Observe the following error is thrown:

![Screenshot 2023-09-05 at 08-57-40 upbeat-meadow-mxljrl - CodeSandbox](https://github.com/firebase/firebase-js-sdk/assets/986290/856be938-d2c8-4c27-8668-f96f6ffc53d0)

I'm assuming it's the same error but `style` has become `w` during minification.

**Comment by prameshj** [comment]
Thanks! I was able to repro with the above link and also on the test app, with a different browser. The draft PR does resolve this error, but i did hit a different assert. I will investigate. Thanks for the inputs!

**Comment by farhaanxxiv** [comment]
It happens with me only for test numbers

**Comment by The-Poe** [comment]
It happens to me in my React project, also in the phone auth and even with real number.

can't just do a exist check before using x.style?
`r && r.style[F] = t`

**Comment by johnfrancmartin** [comment]
We're still running into this, any updates on the Draft PR?

**Comment by andresemartinez** [comment]
I get the same error while resetting a `size: invisible` recaptcha v2 after a firebase error (FirebaseError: Firebase: TOO_SHORT (auth/invalid-phone-number) **only** if the image selector appears. I don't get the error with a `size: normal` recaptcha or if the raptcha verifies without selecting images. The weird part is that is not the actual reset function that breaks, I tried to try catch the error but I couldn't find which function is throwing.

**Comment by Anas-7** [comment]
I have the exact same issue as @andresemartinez where the image selector is causing the issue in the `recaptcha_en.js` file. I face this issue even on normal selectors, and as mentioned when invalid-phone-number error occurs. In my case, I face it even if no error occurs in Firebase sometimes making the entire feature quite unreliable

**Comment by philly-vanilly** [comment]
> HEY GUYS I FOUND THE SOLUTION ...APPARENTLY I WAS USING TEST PHONE NUMBERS BUT ONCE I SIGN IN WITH IT I NEED TO DELETE THAT ACCOUNT FROM FIREBASE USERS SECTION TO BE ABLE TO LOGIN AGAIN WITH THAT TEST NUMBER OTHERWISE RECAPTCHA THROWS THIS ERROR...this worked for me i hope it helps

why are you shouting

**Comment by joeski-porchpass** [comment]
Just wanted to add to this discussion that I opened the above duplicate issue, in which @dlarocque mentioned this open issue last week.

Sorry I missed this thread initially; in my defense, this issue is 5 years old! 😅 

@Anas-7's proposal resolves the issue for us as well, so a big thank you to them! As well as thank you to @dlarocque for pointing us in this direction.

One potentially helpful tidbit to add for fellow Typescript users: this solution calls a private method, so you might have to explicitly tell TypeScript you know what you're doing somewhat similar to this:

```
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/ban-ts-comment */
callback: () => {
  /* @ts-expect-error TS2341 */
  window.recaptchaVerifier?.recaptcha?.reset();
},
"expired-callback": () => {
  /* @ts-expect-error TS2341 */
  window.recaptchaVerifier?.recaptcha?.reset();
},
"error-callback": () => {
  /* @ts-expect-error TS2341 */
  window.recaptchaVerifier?.recaptcha?.reset();
},
/* eslint-enable @typescript-eslint/no-unsafe-call */
/* eslint-enable @typescript-eslint/ban-ts-comment */
```

**Comment by jbalidiong** [comment]
Hello everyone, looking through the last comments, the issue seems to be resolved.

I’ll be closing this issue now. Let me know if there’s any misunderstanding so we can re-open the issue. Or, Please feel free to open a new issue if this still occurs. Thanks!
